### DISCUSSION TITLE
Why is the data not optional in ExecutionResult?

### DISCUSSION THREAD

**Discussion by huihuisang** [maticzav/swift-graphql#134] [Q&A] [answered]
According to the description found in http://spec.graphql.org/October2021/#sec-Response-Format, the data will be null. Therefore, if the data is null, the data.decode process will fail.

<img width="944" alt="image" src="https://github.com/maticzav/swift-graphql/assets/101007168/65e53d05-19c0-46af-a9bb-39940e066ae0">
<img width="1024" alt="image" src="https://github.com/maticzav/swift-graphql/assets/101007168/8ab37dc0-1304-411b-aa4a-7de01a94c7e3">
<img width="937" alt="image" src="https://github.com/maticzav/swift-graphql/assets/101007168/860b5f8f-43af-4489-a199-c725dfb04b33">

**Comment by maticzav** [comment]
The `TypeLock` selection should already account for nullability and `AnyCodable` correctly manages `nil` values.

From the tests and my experience, it should work correctly. Did you find a bug or just curious why it is like that?

**Reply by huihuisang** [reply]
Thank you for replying. @maticzav 

If the data is nil, decoding will fail and throw decoding error, resulting in the loss of information from result.errors.
![](https://user-images.githubusercontent.com/101007168/244565669-860b5f8f-43af-4489-a199-c725dfb04b33.png)

In my case, I receive the JSON data from the server:

`{"errors":[{"message":"Service unavailable for maintenance or that is overloaded. Please try again later.","locations":[{"line":2,"column":3}],"path":["artworksquery__1lu8us4f9flwe"],"extensions":{"code":"INTERNAL_SERVER_ERROR","exception":{"code":50300000,"name":"SERVICE_UNAVAILABLE","expose":false,"statusCode":503,"status":503}}}],"data":null}`

I use `data.decode(selection)` to decode the data, which throws the `SwiftGraphQL.ObjectDecodingError`. Since `result.data` is null, `let data = try selection.decode(raw: result.data)` will immediately trigger a decoding error and lose `result.errors`. Ultimately, I am unable to fetch the error information from the JSON data.

Now it will show an error: `unexpectedObjectType(expected: "Dictionary", received: <null>)`. However, I actually need the message "Service unavailable for maintenance or that is overloaded. Please try again later." in JSON.

**Reply by maticzav** [reply]
Could you also add a snippet of how you are calling the API?

**Reply by huihuisang** [reply]
Code like this:
```
var request = URLRequest(url: url)
request.query(selection, as: operationName)
request.timeoutInterval = 20
request.setValue(defaultUserAgent, forHTTPHeaderField: "User-Agent")
do {
    let (data, response) = try await session.data(for: request)
    switch response.statusCode {
    case 200 ..< 300:
        let result = try data.decode(selection)
        return (result.data, response)
    default:
        …
    }

} catch {
    …
    Log.error(error)
}

```

**Reply by maticzav** [reply]
Thanks, could you add the selection part as well - that's the more important one. 🙂

**Reply by huihuisang** [reply]
It doesn't seem important? I changed to the simplest selection and used it to test with the returned data, but the same error occurred. The selection code like this:
```
let selection =  Selection.Query {
            try $0.media(id: id,
                         selection: Selection.Media {
                             Media(id: try $0.id())
                         }.nullable)
        }
```

**Reply by maticzav** [reply, marked as answer]
@huihuisang thank you for bringing this up and being so patient with me. You are correct that this was indeed an unhandled edge case.

To give you a bit of context, as I described in the PR, I think it's still better to specify `data` as non-nullable because otherwise, we need to modify a lot of code to account for the possibility of `data` being `nil` and that collides with the TypeLocks which resolve `nil` values themself, but I took it for granted that missing fields would work as well.

Thanks!

https://github.com/maticzav/swift-graphql/pull/144

**Reply by huihuisang** [reply]
You're welcome! Thank you for providing the context and explaining your perspective.
May I ask why the new code library has a minimum requirement of only supporting iOS 15? My project currently supports a minimum of iOS 14.😂

**Reply by maticzav** [reply]
https://github.com/maticzav/swift-graphql/issues/136

I know it's not ideal, but I considered that almost no iPhone has iOS 14 anymore (https://www.statista.com/statistics/565270/apple-devices-ios-version-share-worldwide/).

If you gravely need support because you are developing explicitly for iOS 14, I can adjust the parts as a paid feature request 🙂

**Reply by huihuisang** [reply]
Thank you for your response. I understand.
